File: extension/tests/sync/legacyMigration.test.ts
import { describe, expect, test } from "bun:test";
import { activateMigrationAtomically, POMO_BACKUP_V1_WARNING, requireIdentitySelection, validateMigrationInventory, verifyMigrationReady } from "../../src/sync/migration/legacyMigration";

describe("side-by-side legacy migration", () => {
  const verification = { expectedItems: 2, explainedItems: 2, projectionRootsMatch: true, domainInvariantsHold: true };
  const ready = { recoveryAnchorId: "anchor", baselineCaughtUp: true, timerState: "PARKED" as const, identitiesSelected: true, verification };
  test("inventories every durable disposition", () => {
    const dispositions = ["VALID", "DUPLICATE", "CONFLICTING", "MALFORMED_QUARANTINED", "MEMBER_EXCLUDED"] as const;
    validateMigrationInventory({ sourceId: "chrome", durableItemCount: dispositions.length, items: dispositions.map((disposition, index) => ({ sourceId: "chrome", durableId: `id-${index}`, domain: "history", disposition, detail: "classified" })) });
    expect(() => validateMigrationInventory({ sourceId: "chrome", durableItemCount: 2, items: [] })).toThrow(/Every durable/);
  });
  test("blocks identity blending, live timer, stale baseline, and omissions", () => {
    expect(() => requireIdentitySelection([{ memberId: "a", crewIds: new Set(["a"]) }, { memberId: "b", crewIds: new Set(["b"]) }], null)).toThrow(/selection/);
    expect(() => verifyMigrationReady({ ...ready, timerState: "PAUSED" })).toThrow(/Parked/);
    expect(() => verifyMigrationReady({ ...ready, baselineCaughtUp: false })).toThrow(/baseline/);
    expect(() => verifyMigrationReady({ ...ready, verification: { ...verification, explainedItems: 1 } })).toThrow(/omissions/);
  });
  test("atomic activation seals legacy archive and retires dual-write", () => {
    expect(activateMigrationAtomically(ready, "journal", "encrypted-legacy").dualWriteRetired).toBeTrue();
    expect(POMO_BACKUP_V1_WARNING).toContain("sensitive");
  });
});

File: app/src/test/java/com/pomo/sync/migration/LegacyMigrationTest.kt
package com.pomo.sync.migration

import org.junit.Assert.assertTrue
import org.junit.Test

public class LegacyMigrationTest {
    private val ready =
        MigrationPrerequisites(
            "anchor",
            true,
            LegacyTimerState.PARKED,
            true,
            MigrationVerification(2, 2, true, true),
        )

    @Test
    public fun inventoryRequiresDispositionForEveryDurableItem() {
        val kinds =
            LegacyDisposition.entries.mapIndexed { index, kind ->
                LegacyInventoryItem("android", "id-$index", "history", kind, "classified")
            }
        MigrationInventory("android", kinds, kinds.size)
        assertTrue(runCatching { MigrationInventory("android", kinds, kinds.size + 1) }.isFailure)
    }

    @Test
    public fun identityBaselineTimerAndOmissionsBlockCutover() {
        assertTrue(
            runCatching {
                requireIdentitySelection(
                    listOf(MigrationIdentity("a", setOf("crew-a")), MigrationIdentity("b", setOf("crew-b"))),
                    null,
                )
            }.isFailure,
        )
        assertTrue(runCatching { verifyMigrationReady(ready.copy(timerState = LegacyTimerState.PAUSED)) }.isFailure)
        assertTrue(runCatching { verifyMigrationReady(ready.copy(baselineCaughtUp = false)) }.isFailure)
        assertTrue(runCatching { verifyMigrationReady(ready.copy(verification = ready.verification.copy(explainedItems = 1))) }.isFailure)
    }

    @Test
    public fun atomicSuccessSealsArchiveAndPermanentlyRetiresDualWrite() {
        val activation = activateMigrationAtomically(ready, "journal", "encrypted-legacy")
        assertTrue(activation.dualWriteRetired)
        assertTrue(POMO_BACKUP_V1_WARNING.contains("sensitive"))
    }
}
